src/v/storage/disk_log_impl.h

```diff
@@ -36,6 +36,18 @@ class disk_log_impl final : public log::impl {
 public:
     using failure_probes = storage::log_failure_probes;
 
+    /*
+     * the offset index can handle larger than 4gb segments. but the offset
+     * index and compaction index still use 32-bit values to represent some
+     * (relative) logical offsets which mean that a 4-billion limit on records
+     * in a segment exists. by placing a hard limit on segment size we can avoid
+     * those overflows because a segment with more than 4-billion records would
+     * be larger than 4gb even when the records are empty. other practical but
+     * not fundmental limits exist for large segment sizes like our current
+     * in-memory representation of indices.
+     */
+    static constexpr size_t segment_size_hard_limit = 3_GiB;
```

Comment:
If this is intended to be `3_GiB` instead of `4_GiB`, can you name the fiddle factor and subtract it.

Reply:
I agree here w/ @BenPope

Reply:
also, if so, why not do `std::numeric_limits<uint32_t>::max()` instead?

Reply:
what do you want the max to be? I chose 3gb, before there was no max. 3gb gives some wiggle room around the 4gb mark for things like 32 bit relative logical offset limits in the compaction index.

Reply:
@BenPope I think I fixed the fuzz factor issue in the latest force push

Reply:
ok makes sense on the relative things.